influxql/ast.go

```diff
@@ -1012,23 +1020,42 @@ func cloneSource(s Source) Source {
 	}
 }
 
-// RewriteWildcards returns the re-written form of the select statement. Any wildcard query
+// RewriteFields returns the re-written form of the select statement. Any wildcard query
 // fields are replaced with the supplied fields, and any wildcard GROUP BY fields are replaced
-// with the supplied dimensions.
-func (s *SelectStatement) RewriteWildcards(ic IteratorCreator) (*SelectStatement, error) {
+// with the supplied dimensions. Any fields with no type specifier are rewritten with the
+// appropriate type.
+func (s *SelectStatement) RewriteFields(ic IteratorCreator) (*SelectStatement, error) {
+	// Retrieve a list of unique field and dimensions.
+	fieldSet, dimensionSet, err := ic.FieldDimensions(s.Sources)
+	if err != nil {
+		return s, err
+	}
+
+	// Rewrite all variable references in the fields with their types if one
+	// hasn't been specified.
+	rewrite := func(n Node) {
+		if ref, ok := n.(*VarRef); ok {
+			if ref.Type == Unknown || ref.Type == AnyField {
+				if typ, ok := fieldSet[ref.Val]; ok {
+					ref.Type = typ
+				} else if ref.Type != AnyField {
+					if _, ok := dimensionSet[ref.Val]; ok {
+						ref.Type = Tag
+					}
+				}
+			}
+		}
+	}
```

Comment:
This whole function is difficult to read. You can reduce some of the nesting by returning early:

```go
rewrite := func(n Node) {
	if ref, ok := n.(*VarRef); !ok {
		return
	}
	if ref.Type != Unknown && ref.Type != AnyField {
		return
	}

	if typ, ok := fieldSet[ref.Val]; ok {
		ref.Type = typ
	} else if ref.Type != AnyField {
		if _, ok := dimensionSet[ref.Val]; ok {
			ref.Type = Tag
		}
	}
}
```